### PR TITLE
Add undo and redo actions to text blocks

### PR DESCRIPTION
**Pull request by luisbettencourt**
## 🌟 What is the purpose of this PR?

This PR reintroduces the undo and redo actions to text blocks. It was decided that we would implement the native browser undo action for contentEditable elements until the collab rework.

## 🔗 Related links

- [Asana task](https://app.asana.com/0/1200211978612931/1202500568584721/f) _(internal)_

## 🔍 What does this change?

Added the undo and redo actions to prosemirror's keymap.

## ⚠️ Known issues

Both undo and redo actions are very limited and not very consistent. Big chunks of text are grouped together in one step. Also, when selecting a range, overwriting it sometimes breaks functionality not allowing the user to undo or redo those changes. 

## ❓ How to test this?

Play with undo (Cmd + z) and redo (Cmd + y) on any text block.

## 📹 Demo

undo

### CODE REVIEW

File: packages/hash/shared/src/createProseMirrorState.ts [hunk 1]
```diff
@@ -42,6 +42,10 @@ export const createProseMirrorState = ({
       );
       return true;
     },
+    // execCommand is flagged as depecrated but it seems that there isn't a viable alternative
+    // to call the undo and redo default browser actions (https://stackoverflow.com/a/70831583)
```
Comment (nonparibus) [inline packages/hash/shared/src/createProseMirrorState.ts:46]: Please can we inline comment/note that we expect the below code to be removed with forthcoming collab rework?

Comment (luisbettencourt) [reply packages/hash/shared/src/createProseMirrorState.ts:46]: Done

File: packages/hash/shared/src/createProseMirrorState.ts [hunk 2]
```diff
@@ -42,6 +42,11 @@ export const createProseMirrorState = ({
...
+    // After the collab rework this should be replaced with a proper implementation
+    "Mod-z": () => document.execCommand("undo"),
+    "Mod-y": () => document.execCommand("redo"),
```
Comment (kachkaev) [inline packages/hash/shared/src/createProseMirrorState.ts:49]: How about adding ctrl-z / ctrt-shift-z / cmd-shift-z too?

**Comment by nathggns** [comment]
Looking at this now.

File: packages/hash/shared/src/createProseMirrorState.ts (hunk 2 above)
Comment (nathggns) [reply packages/hash/shared/src/createProseMirrorState.ts:49]: I think `ctrl-z` would be covered with `mod` - but would be good to get `mod+shift+z` on here too in order to be a bit more intuitive.

**Review by nathggns — CHANGES_REQUESTED** [review]

File: packages/hash/shared/src/createProseMirrorState.ts (hunk 2 above)
Comment (luisbettencourt) [reply packages/hash/shared/src/createProseMirrorState.ts:49]: Mod-shift-z would is the MacOS shortcut for redo right?

Comment (nathggns) [reply packages/hash/shared/src/createProseMirrorState.ts:49]: Yeah generally I think, though chrome uses cmd+y I think.

Comment (luisbettencourt) [reply packages/hash/shared/src/createProseMirrorState.ts]: Can any of you test if this works of mac please?

File: packages/hash/shared/src/createProseMirrorState.ts [hunk 3]
```diff
@@ -42,6 +42,12 @@ export const createProseMirrorState = ({
       );
       return true;
     },
+    // execCommand is flagged as depecrated but it seems that there isn't a viable alternative
+    // to call the undo and redo default browser actions (https://stackoverflow.com/a/70831583)
+    // After the collab rework this should be replaced with a proper implementation
+    "Mod-z": () => document.execCommand("undo"),
+    "Ctrl-y": () => document.execCommand("redo"),
+    "Cmd-shift-z": () => document.execCommand("redo"),
```
Comment (kachkaev) [inline packages/hash/shared/src/createProseMirrorState.ts:50]:
```suggestion
    "Mod-shift-z": () => document.execCommand("redo"),
```

File: packages/hash/shared/src/createProseMirrorState.ts [hunk 4]
```diff
@@ -42,6 +42,12 @@ export const createProseMirrorState = ({
       );
       return true;
     },
+    // execCommand is flagged as depecrated but it seems that there isn't a viable alternative
+    // to call the undo and redo default browser actions (https://stackoverflow.com/a/70831583)
+    // After the collab rework this should be replaced with a proper implementation
+    "Mod-z": () => document.execCommand("undo"),
+    "Ctrl-y": () => document.execCommand("redo"),
```
Comment (kachkaev) [inline packages/hash/shared/src/createProseMirrorState.ts:49]:
```suggestion
    "Mod-y": () => document.execCommand("redo"),
```

Comment (luisbettencourt) [reply packages/hash/shared/src/createProseMirrorState.ts:49]: Do we want to allow Cmd + y for the redo action in macOs?

**Comment by kachkaev** [comment]
Not sure about the approach in general. It feels pretty unpredictable to me.

Example 1: New block does not prevent undo and pressing Cmd+Z creates a state that did not exist previously


https://user-images.githubusercontent.com/608862/178302371-db33509c-f92b-4f57-8723-e0bccba895dc.mp4



Example 2: Deleting a word with backspace and pressing Cmd+Z afterwards deletes more characters instead of restoring the word


https://user-images.githubusercontent.com/608862/178302389-531d6da4-b225-4744-b542-f1690e859f58.mp4

File: packages/hash/shared/src/createProseMirrorState.ts (hunk 4 above)
Comment (kachkaev) [reply packages/hash/shared/src/createProseMirrorState.ts:49]: Probably yes: https://www.technipages.com/ms-word-undo-redo

> Redo Keyboard Shortcut

(Although I don’t have MS Word to check the article)

**Comment by nonparibus** [comment]
Suggest if Nate gives approval then Joseph can provide functional UAT when he's back, and we can merge for now.

I didn't ask for this and won't be using it in the meantime.

My preference was to wait for collab to do this properly. I don't feel well placed to accept/reject a temporary solution that will be used by @jtewright.

**Comment by luisbettencourt** [comment]
I can also set this PR as a draft until @jtewright returns. Do you prefer to merge it anyway?

**Review by nathggns — APPROVED** [review]
I don't think this makes anything worse than the current situation. This is what was asked for by the person using HASH most right now. I'd merge it.

**Pull request merged**